### DISCUSSION TITLE
Metrics for segmentation

### DISCUSSION THREAD

**Discussion by talhaanwarch** [Lightning-AI/torchmetrics#374] [Q&A] [answered]
Do, we have metrics for segmentation tasks?
I checked [here](https://torchmetrics.readthedocs.io/en/latest/), it only classification and regression

**Comment by SkafteNicki** [comment]
Hi @talhaanwarch,
Sorry for not getting back to you sooner. Which metrics are usually used for segmentation?

**Reply by paudom** [reply]
To mention the most common:
 - DICE Score
 - mIoU (mean Intersection over Union)

**Reply by Sarmadfismael** [reply]
can we use these metrics for sematic segmentation ??

**Comment by Borda** [comment, marked as answer]
@talhaanwarch @paudom @Sarmadfismael — TorchMetrics now has a dedicated segmentation module! Here's what's available (v1.9.0):

**`torchmetrics.segmentation`:**

| Metric | Class |
|--------|-------|
| Dice Score | `DiceScore` |
| Mean IoU | `MeanIoU` |
| Hausdorff Distance | `HausdorffDistance` |
| Generalized Dice | `GeneralizedDiceScore` |
| Panoptic Quality | `PanopticQuality` |

**Quick example — semantic segmentation:**
```python
from torchmetrics.segmentation import DiceScore, MeanIoU

dice = DiceScore(num_classes=21, average="macro", input_format="index")
miou = MeanIoU(num_classes=21, input_format="index")

# preds: (N, H, W) — predicted class indices
# target: (N, H, W) — ground truth
dice_score = dice(preds, target)
miou_score = miou(preds, target)
```

**With Lightning:**
```python
from torchmetrics import MetricCollection
from torchmetrics.segmentation import DiceScore, MeanIoU

seg_metrics = MetricCollection({
    "dice": DiceScore(num_classes=21, average="macro", input_format="index"),
    "miou": MeanIoU(num_classes=21, input_format="index"),
})
self.val_metrics = seg_metrics.clone(prefix="val/")

def validation_step(self, batch, batch_idx):
    preds = self(batch["image"]).argmax(dim=1)
    self.log_dict(self.val_metrics(preds, batch["mask"]))
```

Docs: [DiceScore](https://lightning.ai/docs/torchmetrics/stable/segmentation/dice.html) | [MeanIoU](https://lightning.ai/docs/torchmetrics/stable/segmentation/mean_iou.html)
